engines/scumm/dimuse_v2/dimuse_v2_files.cpp

```diff
@@ -0,0 +1,280 @@
+	int groupId = soundId == kTalkSoundID ? IMUSE_VOLGRP_VOICE : IMUSE_VOLGRP_MUSIC;
+	s = _sound->openSound(soundId, fileName, IMUSE_BUNDLE, groupId, -1);
+	if (!s)
+		s = _sound->openSound(soundId, fileName, IMUSE_BUNDLE, groupId, 1);
+	if (!s)
+		s = _sound->openSound(soundId, fileName, IMUSE_BUNDLE, groupId, 2);
+	if (!s)
+		debug(5, "DiMUSEFilesHandler::openSound(): can't open sound %d (%s)", soundId, fileName);
+}
+
+void DiMUSEFilesHandler::closeSound(int soundId) {
+	_sound->scheduleSoundForDeallocation(soundId);
+}
+
+void DiMUSEFilesHandler::closeAllSounds() {
+	DiMUSESndMgr::SoundDesc *s = _sound->getSounds();
+	for (int i = 0; i < MAX_IMUSE_SOUNDS; i++) {
+		if (s[i].inUse)
+			closeSound(s->soundId);
+	}
+
+	_engine->flushTracks();
+}
+
+void DiMUSEFilesHandler::getFilenameFromSoundId(int soundId, char *fileName) {
+	int i = 0;
+
+	if (soundId == kTalkSoundID) {
+		strcpy(fileName, _currentSpeechFile);
+	}
+
+	if (_vm->_game.id == GID_CMI) {
+		if (soundId < 2000) {
+			while (_comiStateMusicTable[i].soundId != -1) {
+				if (_comiStateMusicTable[i].soundId == soundId) {
+					strcpy(fileName, (char *)_comiStateMusicTable[i].filename);
+					return;
+				}
+				i++;
+			}
+		} else {
+			while (_comiSeqMusicTable[i].soundId != -1) {
+				if (_comiSeqMusicTable[i].soundId == soundId) {
+					strcpy(fileName, (char *)_comiSeqMusicTable[i].filename);
+					return;
+				}
+				i++;
+			}
+		}
+	} else {
+		if (soundId < 2000) {
+			while (_digStateMusicTable[i].soundId != -1) {
+				if (_digStateMusicTable[i].soundId == soundId) {
+					strcpy(fileName, (char *)_digStateMusicTable[i].filename);
+					return;
+				}
+				i++;
+			}
+		} else {
+			while (_digSeqMusicTable[i].soundId != -1) {
+				if (_digSeqMusicTable[i].soundId == soundId) {
+					strcpy(fileName, (char *)_digSeqMusicTable[i].filename);
+					return;
+				}
+				i++;
+			}
+		}
+	}
+}
+
+void DiMUSEFilesHandler::diMUSEAllocSoundBuffer(int bufId, int size, int loadSize, int criticalSize) {
+	DiMUSESoundBuffer *selectedSoundBuf;
+
+	selectedSoundBuf = &_soundBuffers[bufId];
+	selectedSoundBuf->buffer = (uint8 *)malloc(size);
+	selectedSoundBuf->bufSize = size;
+	selectedSoundBuf->loadSize = loadSize;
+	selectedSoundBuf->criticalSize = criticalSize;
+}
+
+void DiMUSEFilesHandler::diMUSEDeallocSoundBuffer(int bufId) {
+	DiMUSESoundBuffer *selectedSoundBuf;
+
+	selectedSoundBuf = &_soundBuffers[bufId];
+	free(selectedSoundBuf->buffer);
+}
+
+void DiMUSEFilesHandler::flushSounds() {
+	DiMUSESndMgr::SoundDesc *s = _sound->getSounds();
+	for (int i = 0; i < MAX_IMUSE_SOUNDS; i++) {
+		DiMUSESndMgr::SoundDesc *curSnd = &s[i];
+		if (curSnd && curSnd->inUse) {
+			if (curSnd->scheduledForDealloc)
+				if (!_engine->diMUSEGetParam(curSnd->soundId, 0x100) && !_engine->diMUSEGetParam(curSnd->soundId, 0x200))
+					_sound->closeSound(curSnd);
+		}
+	}
+}
+
+void DiMUSEFilesHandler::setCurrentSpeechFile(const char *fileName) {
+	strcpy(_currentSpeechFile, fileName);
```

Comment:
buffer size check is missing here.

Reply:
Same as below, let me know if it is acceptable